WebTransport session establishment should handle server trust challenge

AnkshitJain opened the pull request:
#### 8d75fca6f83fdd0cabec51b6140edb0ecf6d0a37
<pre>
WebTransport session establishment should handle server trust challenge
<a href="https://bugs.webkit.org/show_bug.cgi?id=284737">https://bugs.webkit.org/show_bug.cgi?id=284737</a>
<a href="https://rdar.apple.com/136262696">rdar://136262696</a>

Reviewed by Alex Christensen.

Server trust challenges are hooked up to AuthenticationManager.
Tests are updated to use a test delegate to handle server trust challenge.

* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::initializeWebTransportSession):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::initialize): Deleted.
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::didReceiveServerTrustChallenge):
(WebKit::createParameters):
(WebKit::NetworkTransportSession::initialize):
* Source/WebKit/WebProcess/Network/WebSocketProvider.cpp:
(WebKit::WebSocketProvider::initializeWebTransportSession):
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::initialize):
* Source/WebKit/WebProcess/Network/WebTransportSession.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, DISABLED_ClientBidirectional)):
(TestWebKitAPI::TEST(WebTransport, DISABLED_Datagram)):
(TestWebKitAPI::TEST(WebTransport, DISABLED_Unidirectional)):
(TestWebKitAPI::TEST(WebTransport, DISABLED_ServerBidirectional)):

Canonical link: <a href="https://commits.webkit.org/287925@main">https://commits.webkit.org/287925@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/42a39eccf19ccc255b74e2604879fb64f11ade40

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81352 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/877 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35295 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85881 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32338 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83462 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/895 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8694 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63502 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/21260 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84421 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/607 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74037 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43798 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/504 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28201 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30796 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/71994 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28784 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87316 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8582 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71813 "Found 2 new test failures: imported/blink/compositing/animation/hidden-animated-layer-should-not-have-scrollbars.html imported/w3c/web-platform-tests/css/css-animations/animation-offscreen-to-onscreen.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8763 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69863 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71047 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15109 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14013 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12613 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8544 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/14067 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8380 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11901 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10188 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1645,9 +1645,9 @@ void NetworkConnectionToWebProcess::navigatorGetPushPermissionState(URL&& scopeU
 #endif // ENABLE(DECLARATIVE_WEB_PUSH)
 
-void NetworkConnectionToWebProcess::initializeWebTransportSession(URL&& url, WebCore::SecurityOriginData&& origin, CompletionHandler<void(std::optional<WebTransportSessionIdentifier>)>&& completionHandler)
+void NetworkConnectionToWebProcess::initializeWebTransportSession(URL&& url, WebPageProxyIdentifier&& pageID, WebCore::SecurityOriginData&& origin, WebCore::SecurityOriginData&& topOrigin, CompletionHandler<void(std::optional<WebTransportSessionIdentifier>)>&& completionHandler)
```
Let's use a WebCore::ClientOrigin instead of two WebCore::SecurityOriginData so we don't get them mixed up.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -47,20 +50,56 @@
+
+static void didReceiveServerTrustChallenge(Ref<NetworkConnectionToWebProcess>&& connectionToWebProcess, URL&& url, WebKit::WebPageProxyIdentifier&& pageID, WebCore::SecurityOriginData&& topOrigin, sec_trust_t trust, sec_protocol_verify_complete_t completion)
+{
+    const char * host = url.host().utf8().data();
```
nit: no space before *.  Actually, let's not store host in a const char* because any use of it is a UAF because the CString returned by utf8() is a temporary.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -47,20 +50,56 @@
+static void didReceiveServerTrustChallenge(Ref<NetworkConnectionToWebProcess>&& connectionToWebProcess, URL&& url, WebKit::WebPageProxyIdentifier&& pageID, WebCore::SecurityOriginData&& topOrigin, sec_trust_t trust, sec_protocol_verify_complete_t completion)
+{
+    const char * host = url.host().utf8().data();
+    uint16_t port = url.port() ? *url.port() : 443;
```
Let's use defaultPortForProtocol(protocol()) instead of 443.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -47,20 +50,56 @@
+    const char * host = url.host().utf8().data();
+    uint16_t port = url.port() ? *url.port() : 443;
+    RetainPtr secTrust = adoptCF(sec_trust_copy_ref(trust));
+    RetainPtr protectionSpace = adoptNS([[NSURLProtectionSpace alloc] initWithHost:[NSString stringWithUTF8String:host] port:port protocol:NSURLProtectionSpaceHTTPS realm:nil authenticationMethod:NSURLAuthenticationMethodServerTrust]);
```
Yeah, this host use is a UAF.  Let's use initWithHost:url.host().createNSString().get() instead.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -47,20 +50,56 @@
+    uint16_t port = url.port() ? *url.port() : 443;
+    RetainPtr secTrust = adoptCF(sec_trust_copy_ref(trust));
+    RetainPtr protectionSpace = adoptNS([[NSURLProtectionSpace alloc] initWithHost:[NSString stringWithUTF8String:host] port:port protocol:NSURLProtectionSpaceHTTPS realm:nil authenticationMethod:NSURLAuthenticationMethodServerTrust]);
+    [protectionSpace _setServerTrust:secTrust.get()];
```
I'm surprised this compiles without moving _setServerTrust from IPCSerialization.mm to CFNetworkSPI.h.  It probably won't with the non-internal SDK.  Let's do that now.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -47,20 +50,56 @@
+        }
+        RELEASE_ASSERT_NOT_REACHED();
+    };
+    connectionToWebProcess->networkProcess().protectedAuthenticationManager()->didReceiveAuthenticationChallenge(connectionToWebProcess->sessionID(), std::optional { pageID }, &topOrigin, AuthenticationChallenge(challenge.get()), NegotiatedLegacyTLS::No, WTFMove(challengeCompletionHandler));
```
If this compiles without std::optional { } let's do that.  Also the explicit AuthenticationChallenge() might also be unnecessary.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -47,20 +50,56 @@
+                completion(result);
+            }).get());
+            if (status != errSecSuccess)
+                completion(false);
```
This is a use-after-move.  This will always dereference null because completion has been moved from.  In order for something like this to work we need to keep a strong reference to the block and have a strong reference captured in the lambda/block given to SecTrustEvaluateAsyncWithError.

achristensen07 reviewed (changes requested):
One thing I forgot to mention: when we receive a challenge, use the session ID to find a NetworkSessionCocoa, and if sessionCocoa->fastServerTrustEvaluationEnabled() is true, we want to try using NSURLSession._strictTrustEvaluate and complete early if it says it's ok and fall back to the common code you wrote if it doesn't.

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1645,9 +1645,9 @@ void NetworkConnectionToWebProcess::navigatorGetPushPermissionState(URL&& scopeU
 #endif // ENABLE(DECLARATIVE_WEB_PUSH)
 
-void NetworkConnectionToWebProcess::initializeWebTransportSession(URL&& url, WebCore::SecurityOriginData&& origin, CompletionHandler<void(std::optional<WebTransportSessionIdentifier>)>&& completionHandler)
+void NetworkConnectionToWebProcess::initializeWebTransportSession(URL&& url, WebPageProxyIdentifier&& pageID, struct WebCore::ClientOrigin&& clientOrigin, CompletionHandler<void(std::optional<WebTransportSessionIdentifier>)>&& completionHandler)
```
"struct" isn't needed here, though it can be there. Same in header and a few other places in this PR.

achristensen07 commented on Source/WebKit/WebProcess/Network/WebSocketProvider.cpp:
```diff
@@ -47,21 +48,20 @@ RefPtr<ThreadableWebSocketChannel> WebSocketProvider::createWebSocketChannel(Doc
-        ASSERT_NOT_REACHED();
-        return completionHandler(nullptr);
+    if (is<Document>(context)) {
+        if (auto* document = dynamicDowncast<Document>(context)) {
```
RefPtr document

achristensen07 commented on Source/WebKit/WebProcess/Network/WebSocketProvider.cpp:
```diff
@@ -35,6 +35,7 @@
 #include "WebSocketChannelManager.h"
 #include "WebTransportSession.h"
 #include <WebCore/Document.h>
```
nit: this can be removed since DocumentInlines is now included.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -47,20 +50,74 @@
+
+    if (sessionCocoa && sessionCocoa->fastServerTrustEvaluationEnabled()) {
+        auto decisionHandler = makeBlockPtr([sessionCocoa = WeakPtr { *sessionCocoa }, connectionToWebProcess = WTFMove(connectionToWebProcess), pageID = WTFMove(pageID), clientOrigin = WTFMove(clientOrigin), challengeCompletionHandler = WTFMove(challengeCompletionHandler)](NSURLAuthenticationChallenge *challenge, OSStatus trustResult) mutable {
+            auto* session = sessionCocoa.get();
```
We don't need to capture the session or check if it's still alive.  We only need to find it to check fastServerTrustEvaluationEnabled then we never need it again in this flow.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -47,20 +50,74 @@
+static RetainPtr<nw_parameters_t> createParameters(NetworkConnectionToWebProcess& connectionToWebProcess, URL&& url, WebKit::WebPageProxyIdentifier&& pageID, struct WebCore::ClientOrigin&& clientOrigin)
 {
-    auto configureWebTransport = [origin = WTFMove(origin)](nw_protocol_options_t options) {
+    auto configureWebTransport = [clientOrigin = clientOrigin](nw_protocol_options_t options) {
```
clientOrigin = clientOrigin.clientOrigin.toString()  Then we copy less stuff.  And adjust the line below to keep it compiling.

Commit:
WebTransport session establishment should handle server trust challenge
https://bugs.webkit.org/show_bug.cgi?id=284737
rdar://136262696

Reviewed by Alex Christensen.

Server trust challenges are hooked up to AuthenticationManager.
Tests are updated to use a test delegate to handle server trust challenge.

* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::initializeWebTransportSession):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::initialize): Deleted.
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::didReceiveServerTrustChallenge):
(WebKit::createParameters):
(WebKit::NetworkTransportSession::initialize):
* Source/WebKit/WebProcess/Network/WebSocketProvider.cpp:
(WebKit::WebSocketProvider::initializeWebTransportSession):
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::initialize):
* Source/WebKit/WebProcess/Network/WebTransportSession.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, DISABLED_ClientBidirectional)):
(TestWebKitAPI::TEST(WebTransport, DISABLED_Datagram)):
(TestWebKitAPI::TEST(WebTransport, DISABLED_Unidirectional)):
(TestWebKitAPI::TEST(WebTransport, DISABLED_ServerBidirectional)):

Canonical link: https://commits.webkit.org/287925@main
